crates/reth/statediff/src/state.rs

```diff
@@ -0,0 +1,126 @@
+use std::collections::HashMap;
+
+use revm_primitives::{alloy_primitives::Address, B256, U256};
+use strata_mpt::{keccak, MptNode, StateAccount, EMPTY_ROOT};
+
+use crate::BatchStateDiff;
+
+/// An (in-memory) representation of the EVM state reconstructed only from [`BatchStateDiff`].
+#[derive(Clone, Default, Debug)]
+pub struct StateReconstructed {
+    state_trie: MptNode,
+    storage_trie: HashMap<Address, MptNode>,
+}
+
+use thiserror::Error as ThisError;
+
+#[derive(ThisError, Debug)]
+/// A concrete error that may happen during state reconstruction.
+pub enum StateError {
+    #[error("Mpt Error: {0}")]
+    MptError(#[from] strata_mpt::Error),
+}
+
+impl StateReconstructed {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Applies a single [`BatchStateDiff`] atop of the current State.
+    pub fn apply(&mut self, batch_diff: BatchStateDiff) -> Result<(), StateError> {
+        // Insert smart contracts into the state trie.
+        for (address, code) in batch_diff.contracts {
+            // TODO: validate if it's correct all the time?
+            // Are changes to smart contract memory served separately by the slot diffs?
+            let contract_account = StateAccount {
+                nonce: 0,
+                balance: U256::ZERO,
+                storage_root: EMPTY_ROOT,
+                code_hash: code.hash_slow(),
+            };
+            self.state_trie
+                .insert_rlp(&keccak(address), contract_account)?;
+        }
+
+        // Adjust accounts based on the accounts into and its slots.
+        for (address, account_info) in batch_diff.accounts {
+            let acc_info_trie_path = keccak(address);
+            if account_info.is_none() {
+                // Account was actually destructed.
+                self.state_trie.delete(&acc_info_trie_path)?;
+                continue;
+            }
+            let account_info = account_info.unwrap();
```

Comment:
Could write this as

```rs
let Some(account_info) = account_info else {
    // was deleted
    ...
    continue;
};

...
```

Reply:
oh yeah, my c++ roots sometimes show up in different places, but I do agree it's much more rustean way. thanks!